File: src/utils/chaosEngine.ts
/**
 * Chaos Engine Math
 * Real computational formulas for chaos scores, collapse risk, and betting bands
 */

import type { TeamProfile, TotalBand, SpreadBand, ChaosTicketResponse, Game } from '../types';

/**
 * Compute Flop Meter - Late-game meltdown predictor
 * 
 * Measures the risk of a team falling apart or flopping in crucial moments
 * based on fatigue, pace, bench strength, and momentum.
 */
export function computeFlopMeter(
  home: TeamProfile,
  away: TeamProfile,
  game: Game
): number {
  let risk = 0;

  // Pace collapse - slower pace = more tired, more prone to mistakes
  if (game.pace < 92) risk += 10;

  // Star fatigue (OMA teams) - one man army wears down
  if (home.isOma) risk += 8;
  if (away.isOma) risk += 8;

  // Bench scoring drop (Headaches) - can't maintain lead when bench goes cold
  if (home.isHeadaches) risk += 12;
  if (away.isHeadaches) risk += 12;

  // Turnover spike - fatigue causes loose handles
  risk += (home.turnoverRate + away.turnoverRate) / 4;

  // Foul trouble - tired players foul more
  risk += (home.foulRate + away.foulRate) / 4;

  // Momentum swing - large momentum shifts predict collapses
  if (Math.abs(game.momentum) > 40) risk += 10;

  return Math.min(100, Math.max(0, risk));
}

// =========================
// CHAOS ENGINE MATH
// =========================

export function computeChaosScore(home: TeamProfile, away: TeamProfile): number {
  let score = 0;

  // Volatility archetypes
  if (home.isStepperz) score += 15;
  if (away.isStepperz) score += 15;

  if (home.isOma) score += 10;
  if (away.isOma) score += 10;

  if (home.isHeadaches) score += 12;
  if (away.isHeadaches) score += 12;

  // Shooting volatility
  if (home.isBricksquad) score += 8;
  if (away.isBricksquad) score += 8;

  if (home.isHitterz) score -= 5;
  if (away.isHitterz) score -= 5;

  // Recent form
  const gradeMap = { elite: -5, mid: 0, trash: 10 };
  score += gradeMap[home.last3Grade];
  score += gradeMap[away.last3Grade];

  // Quarter volatility
  const qMap = { elite: -3, mid: 0, trash: 5 };
  score += qMap[home.q4Grade];
  score += qMap[away.q4Grade];

  // Risk factors
  score += (home.turnoverRate + away.turnoverRate) / 4;
  score += (home.foulRate + away.foulRate) / 4;

  return Math.min(100, Math.max(0, score));
}

export function computeCollapseRisk(home: TeamProfile, away: TeamProfile): number {
  let risk = 0;

  if (home.isScrubs) risk += 20;
  if (away.isScrubs) risk += 20;

  if (home.isHeadaches) risk += 15;
  if (away.isHeadaches) risk += 15;

  if (away.isFortified) risk += 10; // comeback threat
  if (home.isFortified) risk += 10;

  risk += (home.turnoverRate + away.turnoverRate) / 5;

  return Math.min(100, Math.max(0, risk));
}

export function computeTotalBand(houseTotal: number, chaosScore: number): TotalBand {
  const chaosFactor = chaosScore / 10;

  return {
    low: [houseTotal - (8 + chaosFactor), houseTotal - (4 + chaosFactor)],
    mid: [houseTotal - 3, houseTotal + 3],
    high: [houseTotal + (4 + chaosFactor), houseTotal + (10 + chaosFactor)],
    insane: [houseTotal + (11 + chaosFactor), houseTotal + (22 + chaosFactor)]
  };
}

export function computeSpreadBand(houseSpread: number, home: TeamProfile, away: TeamProfile): SpreadBand {
  let min = houseSpread;
  let max = houseSpread;

  if (home.isScrubs) max += 6;
  if (away.isScrubs) min -= 6;

  if (home.isStepperz) min -= 4;
  if (away.isStepperz) max += 4;

  if (home.isFortified) min -= 3;
  if (away.isFortified) max += 3;

  return { min, max };
}

// =========================
// FULL HOUSE-BAIT ENDPOINT
// =========================

/**
 * Compute House-Bait chaos ticket response
 * 
 * This is the full client-side Chaos Engine function that mirrors the backend.
 * Can be used for testing/demo or offloading computation from backend.
 */
export function computeHouseBait(
  game: Game,
  home: TeamProfile,
  away: TeamProfile,
  payload: {
    houseTotal: number;
    houseSpread: number;
    pickSide: string;
    pickType: "spread" | "ml" | "total";
  }
): ChaosTicketResponse {

  const chaosScore = computeChaosScore(home, away);
  const collapseRisk = computeCollapseRisk(home, away);
  const flopRisk = computeFlopMeter(home, away, game);

  const totalBand = computeTotalBand(payload.houseTotal, chaosScore);
  const spreadBand = computeSpreadBand(payload.houseSpread, home, away);

  // Extract home team archetypes as tags
  const tags = [
    home.isStepperz && "Stepperz",
    home.isScrubs && "Scrubs",
    home.isHeadaches && "Headaches",
    home.isFortified && "Fortified",
    home.isOma && "OMA",
    home.isBricksquad && "Bricksquad",
    home.isHitterz && "Hitterz-Ballerz"
  ].filter(Boolean) as string[];

  // Map tags to human-readable notes
  const notes = tags.map(tag => {
    switch (tag) {
      case "Stepperz": return "Stepperz team: blowout + over-smash risk";
      case "Scrubs": return "Scrubs team: cannot hold leads, fake safety risk";
      case "Headaches": return "Headaches team: slowdown + fake comebacks";
      case "Fortified": return "Fortified: comeback equity from +15";
      case "OMA": return "One Man Army: high volatility when star sits";
      case "Bricksquad": return "Bricksquad: low shooting floor, drought risk";
      case "Hitterz-Ballerz": return "Hitterz-Ballerz: stable shooting floor";
      default: return "";
    }
  }).filter(Boolean);

  return {
    success: true,
    chaosScore,
    collapseRisk,
    flopRisk,
    totalBand,
    spreadBand,
    tags,
    notes
  };
}

File: backend/chaos_engine.py
"""
Chaos Engine Math Module
Real computational formulas for chaos scores, collapse risk, and betting bands.
This is the production Chaos Engine used by the /chaos-ticket endpoint.
"""

from typing import Tuple, Dict, Any
from pydantic import BaseModel


# =========================
# CHAOS ENGINE MATH
# =========================

def compute_chaos_score(home: Dict[str, Any], away: Dict[str, Any]) -> float:
    """
    Compute chaos score based on team archetypes, form, and risk factors.
    
    Args:
        home: Home team profile dict
        away: Away team profile dict
    
    Returns:
        Chaos score 0-100
    """
    score = 0.0

    # Volatility archetypes
    if home.get("isStepperz", False):
        score += 15
    if away.get("isStepperz", False):
        score += 15

    if home.get("isOma", False):
        score += 10
    if away.get("isOma", False):
        score += 10

    if home.get("isHeadaches", False):
        score += 12
    if away.get("isHeadaches", False):
        score += 12

    # Shooting volatility
    if home.get("isBricksquad", False):
        score += 8
    if away.get("isBricksquad", False):
        score += 8

    if home.get("isHitterz", False):
        score -= 5
    if away.get("isHitterz", False):
        score -= 5

    # Recent form
    grade_map = {"elite": -5, "mid": 0, "trash": 10}
    home_last3 = home.get("last3Grade", "mid")
    away_last3 = away.get("last3Grade", "mid")
    score += grade_map.get(home_last3, 0)
    score += grade_map.get(away_last3, 0)

    # Quarter volatility
    q_map = {"elite": -3, "mid": 0, "trash": 5}
    home_q4 = home.get("q4Grade", "mid")
    away_q4 = away.get("q4Grade", "mid")
    score += q_map.get(home_q4, 0)
    score += q_map.get(away_q4, 0)

    # Risk factors
    home_turnovers = home.get("turnoverRate", 0)
    away_turnovers = away.get("turnoverRate", 0)
    home_fouls = home.get("foulRate", 0)
    away_fouls = away.get("foulRate", 0)
    
    score += (home_turnovers + away_turnovers) / 4
    score += (home_fouls + away_fouls) / 4

    return min(100, max(0, score))


def compute_collapse_risk(home: Dict[str, Any], away: Dict[str, Any]) -> float:
    """
    Compute collapse risk based on team volatility and comebacks.
    
    Args:
        home: Home team profile dict
        away: Away team profile dict
    
    Returns:
        Collapse risk 0-100
    """
    risk = 0.0

    if home.get("isScrubs", False):
        risk += 20
    if away.get("isScrubs", False):
        risk += 20

    if home.get("isHeadaches", False):
        risk += 15
    if away.get("isHeadaches", False):
        risk += 15

    if away.get("isFortified", False):
        risk += 10  # comeback threat
    if home.get("isFortified", False):
        risk += 10

    home_turnovers = home.get("turnoverRate", 0)
    away_turnovers = away.get("turnoverRate", 0)
    risk += (home_turnovers + away_turnovers) / 5

    return min(100, max(0, risk))


def compute_flop_meter(home: Dict[str, Any], away: Dict[str, Any], game: Dict[str, Any]) -> float:
    """
    Compute late-game meltdown risk based on pace, fatigue, turnovers, fouls, and momentum.
    
    Args:
        home: Home team profile dict
        away: Away team profile dict
        game: Game state dict with pace, momentum, quarter, timeRemaining
    
    Returns:
        Flop risk 0-100 (late-game meltdown predictor)
    """
    risk = 0.0

    # Pace collapse detection
    pace = game.get("pace", 96)
    if pace < 92:
        risk += 15

    # Star fatigue for OMA teams
    if home.get("isOma", False) or away.get("isOma", False):
        risk += 8

    # Bench scoring drop for Headaches teams
    if home.get("isHeadaches", False) or away.get("isHeadaches", False):
        risk += 12

    # Turnover spike indicates fatigue-driven issues
    home_turnovers = home.get("turnoverRate", 0)
    away_turnovers = away.get("turnoverRate", 0)
    risk += (home_turnovers + away_turnovers) / 4

    # Foul trouble in final quarter
    home_fouls = home.get("foulRate", 0)
    away_fouls = away.get("foulRate", 0)
    risk += (home_fouls + away_fouls) / 4

    # Momentum swing detection
    momentum = game.get("momentum", 0)
    if abs(momentum) > 40:
        risk += 10

    return min(100, max(0, risk))


def compute_total_band(house_total: float, chaos_score: float) -> Dict[str, Tuple[float, float]]:
    """
    Compute total band ranges based on house total and chaos score.
    Higher chaos = wider bands.
    
    Args:
        house_total: House-Bait Total value
        chaos_score: Computed chaos score 0-100
    
    Returns:
        Dict with low, mid, high, insane band ranges as tuples
    """
    chaos_factor = chaos_score / 10

    return {
        "low": (house_total - (8 + chaos_factor), house_total - (4 + chaos_factor)),
        "mid": (house_total - 3, house_total + 3),
        "high": (house_total + (4 + chaos_factor), house_total + (10 + chaos_factor)),
        "insane": (house_total + (11 + chaos_factor), house_total + (22 + chaos_factor))
    }


def compute_spread_band(house_spread: float, home: Dict[str, Any], away: Dict[str, Any]) -> Dict[str, float]:
    """
    Compute spread band ranges based on team archetypes.
    Scrubs widen the spread, Stepperz tighten towards their side, etc.
    
    Args:
        house_spread: House-Bait Spread value
        home: Home team profile dict
        away: Away team profile dict
    
    Returns:
        Dict with min and max spread values
    """
    min_val = house_spread
    max_val = house_spread

    if home.get("isScrubs", False):
        max_val += 6
    if away.get("isScrubs", False):
        min_val -= 6

    if home.get("isStepperz", False):
        min_val -= 4
    if away.get("isStepperz", False):
        max_val += 4

    if home.get("isFortified", False):
        min_val -= 3
    if away.get("isFortified", False):
        max_val += 3

    return {"min": min_val, "max": max_val}


def generate_tags(home: Dict[str, Any], away: Dict[str, Any], chaos_score: float, collapse_risk: float) -> list[str]:
    """
    Generate archetype tags based on team profiles and computed metrics.
    
    Args:
        home: Home team profile dict
        away: Away team profile dict
        chaos_score: Computed chaos score
        collapse_risk: Computed collapse risk
    
    Returns:
        List of archetype tags
    """
    tags = []

    # Check both teams for archetypes
    for team in [home, away]:
        if team.get("isStepperz", False) and "Stepperz" not in tags:
            tags.append("Stepperz")
        if team.get("isScrubs", False) and "Scrubs" not in tags:
            tags.append("Scrubs")
        if team.get("isHeadaches", False) and "Headaches" not in tags:
            tags.append("Headaches")
        if team.get("isFortified", False) and "Fortified" not in tags:
            tags.append("Fortified")
        if team.get("isOma", False) and "OMA" not in tags:
            tags.append("OMA")
        if team.get("isBricksquad", False) and "Bricksquad" not in tags:
            tags.append("Bricksquad")
        if team.get("isHitterz", False) and "Hitterz-Ballerz" not in tags:
            tags.append("Hitterz-Ballerz")

    return tags


def generate_notes(tags: list[str]) -> list[str]:
    """
    Generate human-readable edge notes from archetype tags.
    
    Args:
        tags: List of archetype tags
    
    Returns:
        List of human-readable notes
    """
    note_map = {
        "Stepperz": "Stepperz team detected: blowout + over-smash risk",
        "Scrubs": "Scrubs team: cannot hold leads, fake safety risk",
        "Headaches": "Headaches team: slows down when up, allows fake comebacks",
        "Fortified": "Fortified team: real comeback equity from +15",
        "OMA": "One Man Army: high volatility when star sits",
        "Bricksquad": "Bricksquad: low shooting floor, drought risk",
        "Hitterz-Ballerz": "Hitterz-Ballerz: stable shooting floor"
    }
    
    return [note_map[tag] for tag in tags if tag in note_map]
